File: web-analyst-web/backend/src/domain/attribution/attribution_utils.py
"""
Source attribution utilities for RAG responses.
Addresses mentor feedback: "Could add source attribution in the final answer"
"""
import re
from typing import List, Dict, Tuple
from dataclasses import dataclass


@dataclass
class Citation:
    """Citation information"""
    id: int
    title: str
    url: str
    source_id: str


def format_chunk_id(source_id: str, chunk_index: int) -> str:
    """Format chunk ID for LLM context: [S1:C0]"""
    source_num = ''.join(filter(str.isdigit, source_id)) or '0'
    return f"[S{source_num}:C{chunk_index}]"


def format_chunks_for_llm(chunks: List[Dict]) -> str:
    """
    Format chunks with IDs for LLM context.
    
    Args:
        chunks: List of dicts with 'text' and 'metadata' keys
        
    Returns:
        Formatted string with chunk IDs
    """
    formatted = []
    for chunk in chunks:
        metadata = chunk.get('metadata', {})
        chunk_id = format_chunk_id(
            metadata.get('source_id', '0'),
            metadata.get('chunk_index', 0)
        )
        formatted.append(f"{chunk_id} {chunk['text']}")
    
    return '\n\n'.join(formatted)


def build_citation_map(chunks: List[Dict]) -> Dict[str, Citation]:
    """Build mapping from chunk IDs to citations"""
    citation_map = {}
    source_map = {}
    
    for chunk in chunks:
        metadata = chunk.get('metadata', {})
        source_id = metadata.get('source_id', 'unknown')
        
        if source_id not in source_map:
            source_map[source_id] = {
                'id': len(source_map) + 1,
                'title': metadata.get('title', 'Untitled'),
                'url': metadata.get('url', '')
            }
        
        chunk_id = format_chunk_id(source_id, metadata.get('chunk_index', 0))
        citation_map[chunk_id] = Citation(
            id=source_map[source_id]['id'],
            title=source_map[source_id]['title'],
            url=source_map[source_id]['url'],
            source_id=source_id
        )
    
    return citation_map


def process_attribution(
    llm_response: str,
    chunks: List[Dict]
) -> Tuple[str, List[Citation]]:
    """
    Replace chunk IDs with citation numbers and build citation list.
    
    Args:
        llm_response: LLM response with chunk IDs like [S1:C0]
        chunks: List of chunks used in context
        
    Returns:
        Tuple of (attributed_text, used_citations)
    """
    citation_map = build_citation_map(chunks)
    used_citations = {}
    
    def replace_citation(match):
        chunk_id = match.group(0)
        if chunk_id in citation_map:
            citation = citation_map[chunk_id]
            used_citations[citation.source_id] = citation
            return f"[{citation.id}]"
        return chunk_id
    
    # Replace chunk IDs with citation numbers
    attributed_text = re.sub(r'\[S\d+:C\d+\]', replace_citation, llm_response)
    
    # Build citation list
    citations = sorted(used_citations.values(), key=lambda x: x.id)
    
    if citations:
        citation_list = '\n\nSources:\n' + '\n'.join(
            f"[{c.id}] {c.title}" + (f" — {c.url}" if c.url else "")
            for c in citations
        )
        attributed_text += citation_list
    
    return attributed_text, citations


ATTRIBUTION_SYSTEM_PROMPT = """
When providing your analysis, cite the sources by including the chunk ID in square brackets immediately after the relevant statement.

For example:
- "The research shows significant improvements [S1:C2]."
- "According to the study, the results were positive [S2:C0]."

Always cite the specific chunk that supports your statement. Use the exact chunk ID format provided in the context.
""".strip()

File: web-analyst-web/frontend/src/core/domain/attribution/attributionUtils.ts
export interface ChunkReference {
  sourceId: string;
  chunkIndex: number;
  url?: string;
  title?: string;
}

export interface FormattedChunk {
  id: string;
  text: string;
  reference: ChunkReference;
}

export interface Citation {
  id: number;
  title: string;
  url: string;
  sourceId: string;
}

export function formatChunkId(sourceId: string, chunkIndex: number): string {
  const sourceNum = sourceId.replace(/\D/g, '') || '0';
  return `[S${sourceNum}:C${chunkIndex}]`;
}

export function formatChunksForLLM(chunks: Array<{ text: string; metadata: ChunkReference }>): string {
  return chunks
    .map(chunk => {
      const id = formatChunkId(chunk.metadata.sourceId, chunk.metadata.chunkIndex);
      return `${id} ${chunk.text}`;
    })
    .join('\n\n');
}

export function extractCitationIds(text: string): string[] {
  const regex = /\[S(\d+):C(\d+)\]/g;
  const matches: string[] = [];
  let match;
  while ((match = regex.exec(text)) !== null) {
    matches.push(match[0]);
  }
  return matches;
}

export function buildCitationMap(
  chunks: Array<{ text: string; metadata: ChunkReference }>
): Map<string, Citation> {
  const citationMap = new Map<string, Citation>();
  const sourceMap = new Map<string, { title: string; url: string; count: number }>();

  chunks.forEach(chunk => {
    const { sourceId, url, title } = chunk.metadata;
    if (!sourceMap.has(sourceId)) {
      sourceMap.set(sourceId, {
        title: title || 'Untitled',
        url: url || '',
        count: sourceMap.size + 1,
      });
    }
  });

  chunks.forEach(chunk => {
    const chunkId = formatChunkId(chunk.metadata.sourceId, chunk.metadata.chunkIndex);
    const sourceInfo = sourceMap.get(chunk.metadata.sourceId)!;
    
    citationMap.set(chunkId, {
      id: sourceInfo.count,
      title: sourceInfo.title,
      url: sourceInfo.url,
      sourceId: chunk.metadata.sourceId,
    });
  });

  return citationMap;
}

export function replaceCitationsInText(
  text: string,
  citationMap: Map<string, Citation>
): { text: string; usedCitations: Citation[] } {
  const usedCitationIds = new Set<string>();
  const usedCitations: Citation[] = [];

  const replacedText = text.replace(/\[S(\d+):C(\d+)\]/g, (match) => {
    const citation = citationMap.get(match);
    if (citation) {
      const citationKey = `${citation.sourceId}`;
      if (!usedCitationIds.has(citationKey)) {
        usedCitationIds.add(citationKey);
        usedCitations.push(citation);
      }
      return `[${citation.id}]`;
    }
    return match;
  });

  usedCitations.sort((a, b) => a.id - b.id);

  return { text: replacedText, usedCitations };
}

export function formatCitationList(citations: Citation[]): string {
  if (citations.length === 0) return '';

  return '\n\nSources:\n' + citations
    .map(c => `[${c.id}] ${c.title}${c.url ? ' — ' + c.url : ''}`)
    .join('\n');
}

export function processAnalysisWithAttribution(
  analysisText: string,
  chunks: Array<{ text: string; metadata: ChunkReference }>
): { text: string; citations: Citation[] } {
  const citationMap = buildCitationMap(chunks);
  const { text, usedCitations } = replaceCitationsInText(analysisText, citationMap);
  
  return {
    text: text + formatCitationList(usedCitations),
    citations: usedCitations,
  };
}

export const ATTRIBUTION_SYSTEM_PROMPT = `
When providing your analysis, cite the sources by including the chunk ID in square brackets immediately after the relevant statement.

For example:
- "The research shows significant improvements [S1:C2]."
- "According to the study, the results were positive [S2:C0]."

Always cite the specific chunk that supports your statement. Use the exact chunk ID format provided in the context.
`.trim();
